File: easy/383/383.py
class Solution:
    def canConstruct(self, ransomNote: str, magazine: str) -> bool:
        table = {}
        for n in ransomNote:
            if n not in table:
                table[n] = 1
            else:
                table[n] += 1
        for n in magazine:
            if n in table:
                table[n] -= 1
        for k, v in table.items():
            if v > 0:
                return False
        return True

File: easy/383/383.java
import java.util.HashMap;

class Solution {
    public boolean canConstruct(String ransomNote, String magazine) {
        HashMap<Character, Integer> freq = new HashMap<Character, Integer>();
        for (int i=0;i<magazine.length();i++){
            char c = magazine.charAt(i);
            freq.put(c, freq.getOrDefault(c, 0) + 1);
        }
        for(int j=0;j<ransomNote.length();j++){
            char d = ransomNote.charAt(j);
            freq.put(d, freq.getOrDefault(d, 0) - 1);
            if (freq.get(d) < 0){
                return false;
            }
        }
        return true;
    }
}
// Time complexity O(n + m)
// Space complexity O(k) where k is the length of the unique characters in magazine